// truedata-nodejs/utils.js
module.exports.expires_in = 3600 - 20; // - 20 is to make sure user gets new token 20 seconds before current token expires
module.exports.reconnectionTime = 5000;
module.exports.bidDeactivatedMessage = 'Deactivated';

module.exports.root = 'https://history.truedata.in';
module.exports.login = 'https://auth.truedata.in/token';

module.exports.routes = {
  getTickData: '/getticks',
  getBarData: '/getbars',
  getLastNBars: '/getlastnbars',
  getLastNTicks: '/getlastnticks',
  getBhavCopy: '/getbhavcopy',
  getBhavCopyStatus: '/getbhavcopystatus',
  getTopGainers: '/gettopngainers',
  getTopLosers: '/gettopnlosers',
  getTopVolumeGainers: '/gettopnvolumegainers',
  getCorpAction: '/getcorpaction',
  getSymbolNameChange: '/getsymbolchangehistory',
};

module.exports.formatTime = (year, month, date, hour, min) => {
  return `${year < 100 ? year : year - 2000}${zeroChecker(month)}${zeroChecker(date)}T${zeroChecker(
    hour
  )}:${zeroChecker(min)}:00`;
};

function zeroChecker(number) {
  return number < 10 ? `0${number}` : number;
}

module.exports.zeroChecker = zeroChecker;
